src/node.cc

```diff
@@ -4978,6 +4990,324 @@ int Start(int argc, char** argv) {
+  return context->Global();
+}
+
+
+v8::Local<v8::Value> Call(v8::Local<v8::Object> receiver, v8::Local<v8::Function> function, const std::vector<v8::Local<v8::Value>> & args = {}) {
+    return function->Call(receiver, args.size(), const_cast<v8::Local<v8::Value>*>(&args[0]));
+}
+
+v8::Local<v8::Value> Call(v8::Local<v8::Object> receiver, v8::Local<v8::Function> function, std::initializer_list<v8::Local<v8::Value>> args) {
+    return Call(receiver, function, std::vector<v8::Local<v8::Value>>(args));
+}
+
+// TODO: Error handling: Node.js has exceptions disabled.
+v8::Local<v8::Value> Call(v8::Local<v8::Object> object, const std::string& function_name, const std::vector<v8::Local<v8::Value>>& args) {
+  Local<v8::String> v8_function_name = v8::String::NewFromUtf8(isolate, function_name.c_str());
+
+  Local<v8::Value> value = object->Get(v8_function_name);
+  if (!value->IsFunction()) {
+    //throw new Exception(":((");
+  }
+
+  return Call(object, v8::Local<v8::Function>::Cast(value), args);
+}
+
+v8::Local<v8::Value> Call(v8::Local<v8::Object> object, const std::string & function_name, std::initializer_list<v8::Local<v8::Value>> args) {
+    return Call(object, function_name, std::vector<v8::Local<v8::Value>>(args));
+}
+
+// TODO: Node.js has exceptions disabled.
+v8::Local<v8::Object> IncludeModule(const std::string& module_name) {
+  std::vector<v8::Local<v8::Value>> args = {v8::String::NewFromUtf8(isolate, module_name.c_str())};
+
+  auto module = Call(GetRootObject(), "require", args);
+  if (!module->IsObject()) {
+    // TODO: Can modules not be objects?
+    //throw new Exception(":((");
+  }
+
+  return v8::Local<v8::Object>::Cast(module);
+}
+
+void RegisterModule(const std::string & name, const addon_context_register_func & callback, void *priv) {
+    node::node_module* module = new node::node_module();
+
+    module->nm_version = NODE_MODULE_VERSION;
+    module->nm_flags = NM_F_BUILTIN;
+    module->nm_filename = __FILE__;
+    module->nm_context_register_func = callback;
+    module->nm_modname = name.c_str();
+    module->nm_priv = priv;
+
+    node_module_register(module);
+}
+
+void RegisterModule(const std::string & name,
+                    const std::map<std::string, v8::FunctionCallback> & module_functions) {
+    RegisterModule(name, node::lib::_RegisterModuleCallback, const_cast<std::map<std::string, v8::FunctionCallback>*>(&module_functions));
+}
+
+void _RegisterModuleCallback(v8::Local<v8::Object> exports,
+          v8::Local<v8::Value> module,
+          v8::Local<v8::Context> context,
+          void* priv) {
+    auto module_functions = static_cast<std::map<std::string, v8::FunctionCallback>*>(priv);
+
+    for (std::pair<std::string, v8::FunctionCallback> element : *module_functions) {
+        NODE_SET_METHOD(exports, element.first.c_str(), element.second);
+    }
+}
+
+
+void Terminate() {
+  RequestTerminate();
+  while (ProcessEvents()) { }
+}
+
+void RequestTerminate() {
+  Evaluate("process.exit()");
```

Comment:
This instantly exits the Node.js engine; disregarding any remaining events. 
1) The processEvents() call 4 lines above is unnecessary
2) "RequestTerminate" is possibly the wrong name; this solution is more like a "ForceTerminate"

Reply:
this is not entirely true. If you have listeners on the event 'exit', these functions still get executed. [see here](https://nodejs.org/api/process.html#process_event_exit)